test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -0,0 +1,541 @@
+    std::string  _err_str;
+    ucs_status_t _status;
+};
+
+void UcxCallback::operator()()
+{
+}
+
+UcxContext::UcxContext(size_t iomsg_size) :
+    _listener(NULL), _iomsg_buffer(iomsg_size, '\0')
+{
+    /* Create context */
+    ucp_params_t ucp_params;
+    ucp_params.field_mask   = UCP_PARAM_FIELD_FEATURES |
+                              UCP_PARAM_FIELD_REQUEST_INIT |
+                              UCP_PARAM_FIELD_REQUEST_SIZE;
+    ucp_params.features     = UCP_FEATURE_TAG |
+                              UCP_FEATURE_STREAM;
+    ucp_params.request_init = request_init;
+    ucp_params.request_size = sizeof(ucx_request);
+    ucs_status_t status = ucp_init(&ucp_params, NULL, &_context);
+    if (status != UCS_OK) {
+        throw UcxError("ucp_init", status);
+    }
+
+    log() << "created context " << _context << std::endl;
+
+    /* Create worker */
+    ucp_worker_params_t worker_params;
+    worker_params.field_mask  = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+    status = ucp_worker_create(_context, &worker_params, &_worker);
+    if (status != UCS_OK) {
+        ucp_cleanup(_context);
+        throw UcxError("ucp_worker_create", status);
+    }
+
+    log() << "created worker " << _worker << std::endl;
+
+    post_recv();
+}
+
+UcxContext::~UcxContext()
+{
+    ucp_request_cancel(_worker, _iomsg_recv_request);
+    try {
+        request_wait("iomsg_recv", _iomsg_recv_request);
+    } catch (const UcxError& e) {
+        debug() << "io msg request is completed with status " << e.status()
+                << std::endl;
+    }
+    if (_listener) {
+        ucp_listener_destroy(_listener);
+    }
+    ucp_worker_destroy(_worker);
+    ucp_cleanup(_context);
+    DebugStream::destroy();
+}
+
+void UcxContext::listen(const struct sockaddr* saddr, size_t addrlen)
+{
+    ucp_listener_params_t listener_params;
+
+    listener_params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                                         UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
+    listener_params.sockaddr.addr      = saddr;
+    listener_params.sockaddr.addrlen   = addrlen;
+    listener_params.conn_handler.cb    = connect_callback;
+    listener_params.conn_handler.arg   = reinterpret_cast<void*>(this);
+
+    ucs_status_t status = ucp_listener_create(_worker, &listener_params,
+                                              &_listener);
+    if (status != UCS_OK) {
+        throw UcxError("ucp_listener_create", status);
+    }
+
+    log() << "started listener " << _listener
+          << " on " << sockaddr_str(saddr, addrlen) << std::endl;
+}
+
+UcxConnection* UcxContext::connect(const struct sockaddr* saddr, size_t addrlen)
+{
+    UcxConnection *conn = new UcxConnection(*this, get_next_conn_id(), saddr,
+                                            addrlen);
+    add_connection(conn);
+    return conn;
+}
+
+void UcxContext::disconnect(UcxConnection* conn)
+{
+    // TODO remove from map
```

Comment:
maybe we should remove disconnect() function altogether and use "delete conn" instead?

Reply:
I prefer paired interfaces, create/destroy, connect/disconnect, etc. I'm reworking a bit connection/disconnect flow for better error handling support, so let's see how it will be in another review round.

Reply:
ok